server/src-lib/Hasura/GraphQL/Execute.hs

```diff
@@ -476,173 +457,129 @@ insertBatchResults ::
   -> Batch
   -> GQRespValue
   -> GQRespValue
-insertBatchResults remoteValue batch hasuraResp =
-  case inHashmap (batchRelFieldPath batch) hasuraHash0 remoteHash0 of
+insertBatchResults hasuraResp Batch{..} remoteResp =
+  case inHashmap batchRelFieldPath hasuraData0 remoteData0 of
     Left err  -> appendJoinError hasuraResp (GQJoinError (T.pack err))
     Right val -> GQRespValue (Just (fst val)) (gqRespErrors hasuraResp)
   where
-    hasuraHash0 = fromMaybe OJ.empty (gqRespData hasuraResp)
+    hasuraData0 = fromMaybe OJ.empty (gqRespData hasuraResp)
     -- The 'sortOn' below is not strictly necessary by the spec, but
     -- implementations may not guarantee order of results matching
     -- order of query.
-    remoteHash0 =
-      fromMaybe mempty $ fmap (sortOn fst . OJ.toList) (gqRespData remoteValue)
-    cardinality = biCardinality (batchInputs batch)
+    remoteData0 =
+      maybe mempty (map snd . sortOn fst . OJ.toList) (gqRespData remoteResp)
+      -- ^ TODO I can't figure out what's going on here. It looks like we're
```

Comment:
The remote result is like: 

```
data: {
  remote_result_1 : {},
  remote_result_2: {}
  ...
  remote_result_n: {}
}
```

we want to convert this in the list perserving the order like `[remote_result_1, remote_result_2, ...]`

Reply:
Right, I understand the mechanics of the data transformation there. Part of my question is why should we expect the order of the keys in the result to be sorted to begin with. Presumably we sort the batched query to the remote? If so isn't that very fragile? Again I'm more confused than anything; it might be the function is very sensible but I'd like it to be obvious

Reply:
Yes, this is a very fragile function. We should rewrite it if possible or write a million unit tests. Let's talk on Monday about this join algorithm.